src/pingsocket.rs

```diff
@@ -0,0 +1,179 @@
+        let socket = match d {
+            Domain::IPV4 => Socket::new(Domain::IPV4, Type::RAW, Some(Protocol::ICMPV4))?,
+            Domain::IPV6 => Socket::new(Domain::IPV6, Type::RAW, Some(Protocol::ICMPV6))?,
+            _ => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidData,
+                    "Invalid domain",
+                ))
+            }
+        };
+
+        // TODO: Type filtering,
+        // https://tools.ietf.org/html/rfc3542#section-3.2. Currently blocked
+        // on https://github.com/rust-lang/socket2/issues/199
+
+        // TODO: Get access to the hop limits
+        // https://tools.ietf.org/html/rfc3542#section-4, to show the TTL for
+        // ICMPv6.
+        socket.set_nonblocking(true)?;
+        Ok(PingSocketBuilder { socket })
+    }
+    #[cfg(any(target_os = "android", target_os = "fuchsia", target_os = "linux"))]
+    pub fn bind_device(&self, interface: Option<&[u8]>) -> io::Result<()> {
+        self.socket.bind_device(interface)
+    }
+
+    #[cfg(all(feature = "all", any(target_os = "freebsd")))]
+    pub fn set_fib(&self, fib:u32) -> io::Result<()> {
+        self.socket.set_fib(fib)
+    }
+
+    pub fn bind_addr(&self, sock_addr: &SockAddr) -> io::Result<()> {
+        self.socket.bind(sock_addr)
+    }
+
+    pub fn set_ttl(&self, ttl: u32) -> io::Result<()> {
+        self.socket.set_ttl(ttl)
+    }
+
+    pub fn set_send_buffer_size(&self, bufsize: usize) -> io::Result<()> {
+        self.socket.set_send_buffer_size(bufsize)
+    }
+
+    pub fn set_recv_buffer_size(&self, bufsize: usize) -> io::Result<()> {
+        self.socket.set_recv_buffer_size(bufsize)
+    }
+    fn inner_run(self) -> io::Result<UdpSocket> {
+        #[cfg(windows)]
+        return Ok(UdpSocket::from_std(unsafe {
+            std::net::UdpSocket::from_raw_socket(self.socket.into_raw_socket())
+        })?);
+        #[cfg(unix)]
+        return Ok(UdpSocket::from_std(unsafe {
+            std::net::UdpSocket::from_raw_fd(self.socket.into_raw_fd())
+        })?);
+    }
+
+    pub fn run(self) -> io::Result<PingSocket> {
+        PingSocket::new_socket(AsyncSocket::new(self.inner_run()?))
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct AsyncSocket {
+    inner: Arc<UdpSocket>,
+}
+impl AsyncSocket {
+    fn new(socket: UdpSocket) -> Self {
+        AsyncSocket {
+            inner: Arc::new(socket),
+        }
+    }
+    pub async fn recv_from(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)> {
+        self.inner.recv_from(buf).await
+    }
+    pub async fn send_to(&self, buf: &mut [u8], target: &SocketAddr) -> io::Result<usize> {
+        self.inner.send_to(buf, target).await
+    }
+}
+#[derive(Clone)]
+pub struct PingSocket {
+    inner: AsyncSocket,
+    pmap: Arc<Mutex<BTreeMap<IpAddr, Sender<Vec<u8>>>>>,
+    recv_task: Arc<Mutex<Option<tokio::task::JoinHandle<()>>>>,
+}
+
+impl PingSocket {
+    pub fn new(d: Domain) -> io::Result<PingSocket> {
```

Comment:
Is this for compatibility with the old interface? But it feels that the design is a bit cumbersome. If you simplify the usage, we can upgrade to a larger version.

Reply:
But it will broken compatibility.
I think it acceptable on major version change.
If you think we are ready to this - it is a reason to upgrade API.

Reply:
BTW, now I'm testing this crate with set_fib method from socket2, which only available yet on github, not on crates.io, so dependency in Cargo.toml replaced.
I think it need to be reversed before merge.
When you will be ready to merge, I rollback this change.

Reply:
ok, please eliminate some warnings prompted by `rust-clippy`, and execute `rustfmt`. I will do the merge later.

Reply:
done